packages/firestore/src/remote/online_state_tracker.ts

```diff
@@ -0,0 +1,166 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { OnlineState } from '../core/types';
+import * as log from '../util/log';
+import { assert } from '../util/assert';
+import { AsyncQueue, TimerId } from '../util/async_queue';
+import { CancelablePromise } from '../util/promise';
+
+const LOG_TAG = 'OnlineStateTracker';
+
+// To deal with transient failures, we allow multiple stream attempts before
+// giving up and transitioning to Offline.
+const MAX_WATCH_STREAM_FAILURES = 2;
+
+// To deal with stream attempts that don't succeed or fail in a timely manner,
+// we have a maximum timeout we'll wait for the stream to either succeed or fail
+// MAX_WATCH_STREAM_FAILURES times, else we revert to OnlineState.Offline.
+const MAX_WATCH_STREAM_TIMEOUT_MS = 10 * 1000;
+
+/**
+ * A component used by the RemoteStore to track the OnlineState (that is,
+ * whether or not the client as a whole should be considered to be online or
+ * offline), implementing the appropriate heuristics.
+ *
+ * In particular, when the client is trying to connect to the backend, we
+ * allow up to MAX_WATCH_STREAM_FAILURES within MAX_WATCH_STREAM_TIMEOUT_MS for
+ * a connection to succeed. If we have too many failures or the timeout elapses,
+ * then we set the OnlineState to Offline, and the client will behave as if
+ * it is offline (get()s will return cached data, etc.).
+ */
+export class OnlineStateTracker {
+  /** The current OnlineState. */
+  private state = OnlineState.Unknown;
+
+  /**
+   * A count of consecutive failures to open the stream. If it reaches the
+   * maximum defined by MAX_WATCH_STREAM_FAILURES, we'll set the OnlineState to
+   * Offline.
+   */
+  private watchStreamFailures = 0;
+
+  /**
+   * A timer that elapses after MAX_WATCH_STREAM_TIMEOUT_MS, at which point we
+   * revert to OnlineState.Offline without waiting for the stream to actually
+   * fail (MAX_WATCH_STREAM_FAILURES times).
+   */
+  private watchStreamTimerPromise: CancelablePromise<void> | null = null;
+
+  /**
+   * Whether the client should log a warning message if it fails to connect to
+   * the backend (initially true, cleared after a successful stream, or if we've
+   * logged the message already).
+   */
+  private shouldWarnClientIsOffline = true;
+
+  constructor(
+    private asyncQueue: AsyncQueue,
+    private onlineStateHandler: (onlineState: OnlineState) => void
+  ) {}
+
+  /**
+   * Called by RemoteStore when a watch stream is started.
+   *
+   * It sets the OnlineState to Unknown and starts a MAX_WATCH_STREAM_TIMEOUT_MS
+   * timer if necessary.
+   */
+  handleWatchStreamStart(): void {
+    this.setAndBroadcast(OnlineState.Unknown);
+
+    if (this.watchStreamTimerPromise === null) {
+      this.watchStreamTimerPromise = this.asyncQueue.enqueueAfterDelay(
+        TimerId.OnlineStateTimeout,
```

Comment:
I find this cluster of names confusingly dissimlar. Why is this an "OnlineStateTimeout", the promise is a "watchStreamTimerPromise", and the timeout a "MAX_WATCH_STREAM_TIMEOUT_MS"?

These all seem related to the handling the initial connection but:
  * OnlineStateTimeout seems like it might be applicable to other transitions
  * watchStreamTimer seems wholly unrelated to that purpose
  * MAX_WATCH_STREAM_TIMEOUT_MS seems just as likely to apply to an idle timeout. Also, why "MAX"? Isn't this just the timeout duration?

Reply:
This is a marked improvement and LGTM as-is. Thanks. Agree that leaning on comments can be the way to go in cases like these.

I'm still torn about this because the connection timer seems very much related to what we're doing here: namely if we're ever connecting our state starts at unknown. If we logically fail to connect (i.e. fail to transition out of unknown) then this timer fires. It seems like, for all intents and purposes, OnlineStateUnknownTimeout is the real connection timeout. Could that other thing we have be some lower-level concept?

Or maybe the name of OnlineStateUnknownTimeout itself is caught in my craw. Our other timeouts are named in terms of the end state, not the starting state. IOW: how long until some state, not how long in some state. How long until idle, how long until connected, etc.

This comment isn't super actionable but I'm still uneasy, and if history has shown anything: this code is fiddly, keeps changing, and we keep misunderstanding it over time, so if you have the energy I'd like to try to find a way to make it more immediately obvious and less fuzzy.

Reply:
I agree that our timers have been fiddly / confusing and getting this right is worthwhile. :-)

So right now our timers are: ListenStreamIdle / ListenStreamConnection, WriteStreamIdle / WriteStreamConnection, OnlineStateTimeout

The Connection timers are used to schedule delayed stream attempts subject to exponential backoff...  Not all stream attempts are delayed in which case the "Connection" timer doesn't come into play (e.g the first attempt once a stream becomes necessary, or the first attempt after an established stream encounters an error, etc.). 

But the OnlineStateTimeout is still relevant when the stream attempt isn't delayed (e.g. it'll fire if the first stream attempt doesn't succeed / fail within 10 seconds, triggering us to OnlineState.Offline while we wait).

And the Connection timers will continue to be used even once the OnlineStateTimeout has triggered (because we'll continue scheduling stream attempts with backoff even after we've entered OnlineState.Offline).  

So I think they're not actually that related, and we probably have a naming problem that makes them seem related. We could try to be more verbose, e.g. ListenStreamDelayedRestart instead of ListenStreamConnection.

In naming these, it's worth thinking about their usage from tests, e.g.: runDelayedOperationsUntil(TimerId.ListenStreamDelayedRestart) to make a test skip the exponential backoff delay.

Reply:
OK could the connection timer ids really be backoff timer ids? This is essentially where we use them:

```js
    this.backoff = new ExponentialBackoff(
      queue,
      connectionTimerId,
      BACKOFF_INITIAL_DELAY_MS,
      BACKOFF_FACTOR,
      BACKOFF_MAX_DELAY_MS
    );
```

`TimerId.ListenStreamBackoff` => `runDelayedOperationsUntil(TimerId.ListenStreamBackoff)`?

I don't think the fact that the online state timeout is for succeeding or failing much matters. Connection timeouts pretty much always work this way: they introduce a client-generated error if the connection hasn't already succeeded or failed on its own.

Reply:
Yes, that's exactly what they are, except "backoff" describes an implementation detail about how the timers are scheduled rather than what they actually are for. You could hypothetically imagine that the stream could have multiple timers, for different purposes, that happen to use exponential backoff scheduling, at which point "ListenStreamBackoff" would be meaningless (as opposed to "ListenStreamDelayedRestart" or whatever).

That said, we probably conflate "backoff" with the act of reconnecting already, and so I'd be okay with ListenStreamBackoff if you think that's clearer.

Reply:
Also, I still do think there's a meaningful distinction between our "online state timeout" and a "connection timeout" because we explicitly don't cancel the existing stream attempt and start a new one...  Instead we're relying on grpc's internal timeout / retry connection logic.